AWSAuthSDK/Sources/AWSMobileClient/AWSAsyncBlockOperation.swift

```diff
@@ -0,0 +1,70 @@
+import Foundation
+
+enum AWSAsyncBlockState: String {
+    case ready = "isReady"
+    case executing = "isExecuting"
+    case finished  = "isFinished"
+    case cancelled = "isCancelled"
+}
+
+public typealias AWSAsyncBlockDoneClosure = () -> Void
+public typealias AWSAsyncBlockClosure = (@escaping AWSAsyncBlockDoneClosure) -> Void
+
+public class AWSAsyncBlockOperation: Operation {
```

Comment:
This looks similar to https://github.com/aws-amplify/amplify-ios/blob/main/Amplify/Core/Support/AsychronousOperation.swift which was introduced in Amplify iOS for operations that require to do some async work. I'm thinking
- If the intent of the two classes are pretty much the same, can we review `AsynchronousOperation` with your learnings and identify if there are any changes we should do to the one in Amplify? Not blocking this PR, just can keep an issue open for the changes we want to make to consolate the two in terms of behavior and usability, as we want to avoid breaking changes unless it's really needed (performance improvement, incorrect usage of subclassing Operation)
- Should `AWSAsyncBlockOperation` be in AWSCore with some unit tests? I'm guessing this can be reused in other places like TransferUtility or Pinpoint SDK if the same code patterns were applied there

Reply:
We discussed this during a meeting recently. This type is similar but is also in AWS Mobile Client and built in Swift code. AWS Core is Objective-C and this module is a dependency of Amplify Auth. AsynchronousOperation also does a bit more than this type does which includes integration with other Amplify systems including Combine features. AsyncBlockOperation does as little as possible to integrate properly with an `OperationQueue` based on Apple's docs to eliminate the need for any blocking to free up threads. The 2 key mechanisms are the KVO notifications and the `done` block which moves to the finished state so the next `Operation` can start.